<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Exploring-Weather-Trends---Project-Instructions" data-toc-modified-id="Exploring-Weather-Trends---Project-Instructions-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Exploring Weather Trends - Project Instructions</a></span><ul class="toc-item"><li><span><a href="#Summary" data-toc-modified-id="Summary-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Summary</a></span></li><li><span><a href="#Instructions" data-toc-modified-id="Instructions-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Instructions</a></span></li><li><span><a href="#Submission" data-toc-modified-id="Submission-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Submission</a></span></li><li><span><a href="#Rubric" data-toc-modified-id="Rubric-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Rubric</a></span></li></ul></li><li><span><a href="#Extract-the-data" data-toc-modified-id="Extract-the-data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Extract the data</a></span><ul class="toc-item"><li><span><a href="#Write-a-SQL-query-to-extract-the-city-level-data.-Export-to-CSV." data-toc-modified-id="Write-a-SQL-query-to-extract-the-city-level-data.-Export-to-CSV.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Write a SQL query to extract the city level data. Export to CSV.</a></span></li><li><span><a href="#Write-a-SQL-query-to-extract-the-global-data.-Export-to-CSV." data-toc-modified-id="Write-a-SQL-query-to-extract-the-global-data.-Export-to-CSV.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Write a SQL query to extract the global data. Export to CSV.</a></span></li></ul></li><li><span><a href="#Open-up-the-CSV" data-toc-modified-id="Open-up-the-CSV-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Open up the CSV</a></span><ul class="toc-item"><li><span><a href="#guarulhos_city_data.csv" data-toc-modified-id="guarulhos_city_data.csv-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>guarulhos_city_data.csv</a></span></li><li><span><a href="#global_data.csv" data-toc-modified-id="global_data.csv-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>global_data.csv</a></span></li></ul></li><li><span><a href="#Create-a-line-chart" data-toc-modified-id="Create-a-line-chart-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Create a line chart</a></span><ul class="toc-item"><li><span><a href="#City-temperatures" data-toc-modified-id="City-temperatures-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>City temperatures</a></span></li><li><span><a href="#Global-temperatures" data-toc-modified-id="Global-temperatures-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Global temperatures</a></span></li></ul></li><li><span><a href="#Make-observations" data-toc-modified-id="Make-observations-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Make observations</a></span><ul class="toc-item"><li><span><a href="#Is-your-city-hotter-or-cooler-on-average-compared-to-the-global-average?-Has-the-difference-been-consistent-over-time?" data-toc-modified-id="Is-your-city-hotter-or-cooler-on-average-compared-to-the-global-average?-Has-the-difference-been-consistent-over-time?-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Is your city hotter or cooler on average compared to the global average? Has the difference been consistent over time?</a></span></li><li><span><a href="#How-do-the-changes-in-your-city’s-temperatures-over-time-compare-to-the-changes-in-the-global-average?" data-toc-modified-id="How-do-the-changes-in-your-city’s-temperatures-over-time-compare-to-the-changes-in-the-global-average?-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>How do the changes in your city’s temperatures over time compare to the changes in the global average?</a></span></li><li><span><a href="#What-does-the-overall-trend-look-like?-Is-the-world-getting-hotter-or-cooler?-Has-the-trend-been-consistent-over-the-last-few-hundred-years?" data-toc-modified-id="What-does-the-overall-trend-look-like?-Is-the-world-getting-hotter-or-cooler?-Has-the-trend-been-consistent-over-the-last-few-hundred-years?-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>What does the overall trend look like? Is the world getting hotter or cooler? Has the trend been consistent over the last few hundred years?</a></span></li></ul></li><li><span><a href="#Steps-taken-to-prepare-the-data-to-be-visualized-in-the-chart" data-toc-modified-id="Steps-taken-to-prepare-the-data-to-be-visualized-in-the-chart-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Steps taken to prepare the data to be visualized in the chart</a></span><ul class="toc-item"><li><span><a href="#What-tools-did-you-use-for-each-step?-(Python,-SQL,-Excel,-etc)" data-toc-modified-id="What-tools-did-you-use-for-each-step?-(Python,-SQL,-Excel,-etc)-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>What tools did you use for each step? (Python, SQL, Excel, etc)</a></span></li><li><span><a href="#How-did-you-calculate-the-moving-average?" data-toc-modified-id="How-did-you-calculate-the-moving-average?-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>How did you calculate the moving average?</a></span></li><li><span><a href="#What-were-your-key-considerations-when-deciding-how-to-visualize-the-trends?" data-toc-modified-id="What-were-your-key-considerations-when-deciding-how-to-visualize-the-trends?-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>What were your key considerations when deciding how to visualize the trends?</a></span></li></ul></li><li><span><a href="#Suggestions-to-Make-Your-Project-Stand-Out" data-toc-modified-id="Suggestions-to-Make-Your-Project-Stand-Out-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Suggestions to Make Your Project Stand Out</a></span><ul class="toc-item"><li><span><a href="#What's-the-correlation-coefficient?" data-toc-modified-id="What's-the-correlation-coefficient?-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>What's the correlation coefficient?</a></span></li><li><span><a href="#Can-you-estimate-the-average-temperature-in-your-city-based-on-the-average-global-temperature?" data-toc-modified-id="Can-you-estimate-the-average-temperature-in-your-city-based-on-the-average-global-temperature?-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Can you estimate the average temperature in your city based on the average global temperature?</a></span></li><li><span><a href="#Multiple-cities---Add-your-favorite-cities-from-around-the-globe-to-your-visualization.-What-do-you-learn-about-them?" data-toc-modified-id="Multiple-cities---Add-your-favorite-cities-from-around-the-globe-to-your-visualization.-What-do-you-learn-about-them?-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Multiple cities - Add your favorite cities from around the globe to your visualization. What do you learn about them?</a></span></li></ul></li></ul></div>

In [1]:
import pandas as pd

# Exploring Weather Trends - Project Instructions

## Summary
In this project, you will analyze local and global temperature data and compare the temperature trends where you live to overall global temperature trends.

## Instructions
Your goal will be to create a visualization and prepare a write up describing the similarities and differences between global temperature trends and temperature trends in the closest big city to where you live. To do this, you’ll follow the steps below:
* **Extract the data** from the database. There's a workspace in the next section that is connected to a database. You’ll need to export the temperature data for the world as well as for the closest big city to where you live. You can find a list of cities and countries in the `city_list` table. To interact with the database, you'll need to write a SQL query.
 * Write a SQL query to extract the city level data. Export to CSV.
 * Write a SQL query to extract the global data. Export to CSV.
* **Open up the CSV** in whatever tool you feel most comfortable using. We suggest using Excel or Google sheets, but you are welcome to use another tool, such as Python or R.
* **Create a line chart** that compares your city’s temperatures with the global temperatures. Make sure to plot the _moving average_ rather than the yearly averages in order to smooth out the lines, making trends more observable (the last concept in the previous lesson goes over how to do this in a spreadsheet).
* **Make observations** about the similarities and differences between the world averages and your city’s averages, as well as overall trends. Here are some questions to get you started.
 * Is your city hotter or cooler on average compared to the global average? Has the difference been consistent over time?
 * “How do the changes in your city’s temperatures over time compare to the changes in the global average?”
 * What does the overall trend look like? Is the world getting hotter or cooler? Has the trend been consistent over the last few hundred years?
 
## Submission
Your submission should be a PDF that includes:
* **An outline** of steps taken to prepare the data to be visualized in the chart, such as:
 * What tools did you use for each step? (Python, SQL, Excel, etc)
 * How did you calculate the moving average?
 * What were your key considerations when deciding how to visualize the trends?
* **Line chart** with local and global temperature trends
* At least **four observations** about the similarities and/or differences in the trends

## Rubric
A Udacity reviewer will assess your project based on the criteria in the project rubric. Use the rubric as a guide while you complete the project, then give yourself a quick self-assessment before you submit it.

If you're on this page, then you should have completed your exploration of local and global temperature trends. Congratulations! Before you submit your project, make sure to check the following points:

* Please submit your project as a PDF. Your report should include documentation of the steps in your analysis, a line chart depicting the local and global temperature data, and your observations regarding the trends.
* Don't forget to review the rubric! Reviewers will use the rubric to assess your work, so make sure it 'meets specifications' on all points before you submit.

Once you've checked the above, click on the "Submit Project" button below to go to the project submission page. After you submit your project, it can take up to a week for it to be evaluated. Most of the time, it is much faster! In the meantime, you can feel free to continue to other parts of the program to continue your learning!

# Extract the data

## Write a SQL query to extract the city level data. Export to CSV.

```SQL
-- Average temperatures for Guarulhos city by year (ºC).
SELECT * FROM city_data
  WHERE city = 'Guarulhos'
  AND avg_temp IS NOT NULL;
```

## Write a SQL query to extract the global data. Export to CSV.

```SQL
-- Average global temperatures by year (ºC).
SELECT * FROM global_data;
```

# Open up the CSV

## guarulhos_city_data.csv

In [2]:
guarulhos_city_data = pd.read_csv("./data/guarulhos_city_data.csv")
print(guarulhos_city_data)

     year       city country  avg_temp
0    1832  Guarulhos  Brazil     18.94
1    1833  Guarulhos  Brazil     19.93
2    1834  Guarulhos  Brazil     19.21
3    1835  Guarulhos  Brazil     18.62
4    1836  Guarulhos  Brazil     18.84
5    1837  Guarulhos  Brazil     18.28
6    1838  Guarulhos  Brazil     18.72
7    1839  Guarulhos  Brazil     18.52
8    1840  Guarulhos  Brazil     19.20
9    1841  Guarulhos  Brazil     18.89
10   1842  Guarulhos  Brazil     19.32
11   1843  Guarulhos  Brazil     19.46
12   1851  Guarulhos  Brazil     19.44
13   1852  Guarulhos  Brazil     19.60
14   1853  Guarulhos  Brazil     19.61
15   1854  Guarulhos  Brazil     19.70
16   1855  Guarulhos  Brazil     19.70
17   1856  Guarulhos  Brazil     18.79
18   1857  Guarulhos  Brazil     19.39
19   1858  Guarulhos  Brazil     18.39
20   1859  Guarulhos  Brazil     19.05
21   1860  Guarulhos  Brazil     19.73
22   1861  Guarulhos  Brazil     18.99
23   1862  Guarulhos  Brazil     19.01
24   1863  Guarulhos  Bra

## global_data.csv

In [3]:
global_data = pd.read_csv("./data/global_data.csv")
print(global_data)

     year  avg_temp
0    1750      8.72
1    1751      7.98
2    1752      5.78
3    1753      8.39
4    1754      8.47
5    1755      8.36
6    1756      8.85
7    1757      9.02
8    1758      6.74
9    1759      7.99
10   1760      7.19
11   1761      8.77
12   1762      8.61
13   1763      7.50
14   1764      8.40
15   1765      8.25
16   1766      8.41
17   1767      8.22
18   1768      6.78
19   1769      7.69
20   1770      7.69
21   1771      7.85
22   1772      8.19
23   1773      8.22
24   1774      8.77
25   1775      9.18
26   1776      8.30
27   1777      8.26
28   1778      8.54
29   1779      8.98
..    ...       ...
236  1986      8.83
237  1987      8.99
238  1988      9.20
239  1989      8.92
240  1990      9.23
241  1991      9.18
242  1992      8.84
243  1993      8.87
244  1994      9.04
245  1995      9.35
246  1996      9.04
247  1997      9.20
248  1998      9.52
249  1999      9.29
250  2000      9.20
251  2001      9.41
252  2002      9.57
253  2003      9.53


# Create a line chart

## City temperatures

![Image](./image/guarulhos_city_data_year_c.png)

## Global temperatures

![Image](./image/global_data_year_c.png)

# Make observations

## Is your city hotter or cooler on average compared to the global average? Has the difference been consistent over time?

Guarulhos city temperature average is hotter compared to the global average. Yes, the slope line seems very similar in two plots.

## How do the changes in your city’s temperatures over time compare to the changes in the global average?

In my city the temperature average is not much linear, with more fluctuations in same period compared to the global average.

## What does the overall trend look like? Is the world getting hotter or cooler? Has the trend been consistent over the last few hundred years?

In the graph we see the line approaching to the 10°C. The world is getting hotter.
![Image](./image/global_data_few_hundred_year_c.png)
In the last few hundred years we see the line approaching to the 10°C, it is consistent to the overall trend looks like.

# Steps taken to prepare the data to be visualized in the chart

## What tools did you use for each step? (Python, SQL, Excel, etc)

Google Sheets. I opened each csv file in separate document, than I create a new document and copy all previous document data into new document in separate sheets. I plotted the graph into a new sheet of this document.

## How did you calculate the moving average?

I used the `AVERAGE()` function of Google Sheets selecting the previous 10 cells of column `avg_temp`, corresponding the 10 past years registred. I clicked and dragged the formula down to the next cell until reach the last one.

## What were your key considerations when deciding how to visualize the trends?

I removed all null `avg_temp` field value to extract `CSV` file of `city_data` table. To plot graph I considered the same minimum and maximum year to show the same period. I scaled y-axes to same range of temperature, so it would be easier to see the distance between two lines.

# Suggestions to Make Your Project Stand Out

## What's the correlation coefficient?

Is a measure of the linear correlation between two variables.  it has a value between +1 and −1, where 1 is total positive linear correlation, 0 is no linear correlation, and −1 is total negative linear correlation. It is widely used in the sciences.

>source: https://en.wikipedia.org/wiki/Pearson_correlation_coefficient

## Can you estimate the average temperature in your city based on the average global temperature?

**Data extraction**
```SQL
  SELECT cd.year
       , cd.avg_temp AS cd_avg_temp
       , gd.avg_temp AS gd_avg_temp
    FROM global_data AS gd
    JOIN city_data AS cd
      ON cd.year = gd.year
   WHERE cd.avg_temp IS NOT NULL
     AND cd.city = 'Guarulhos';
```

**Save data into `guarulhos_city_global_data.csv` file**

In [4]:
guarulhos_city_global_data = pd.read_csv("./data/guarulhos_city_global_data.csv")
print(guarulhos_city_global_data)

     cd_year  cd_avg_temp  gd_avg_temp
0       1832        18.94         7.45
1       1833        19.93         8.01
2       1834        19.21         8.15
3       1835        18.62         7.39
4       1836        18.84         7.70
5       1837        18.28         7.38
6       1838        18.72         7.51
7       1839        18.52         7.63
8       1840        19.20         7.80
9       1841        18.89         7.69
10      1842        19.32         8.02
11      1843        19.46         8.17
12      1851        19.44         8.18
13      1852        19.60         8.10
14      1853        19.61         8.04
15      1854        19.70         8.21
16      1855        19.70         8.11
17      1856        18.79         8.00
18      1857        19.39         7.76
19      1858        18.39         8.10
20      1859        19.05         8.25
21      1860        19.73         7.96
22      1861        18.99         7.85
23      1862        19.01         7.56
24      1863        18.91

**Import `CSV` file into Google Sheet and use `correl()` function using `cd_avg_temp` and `gd_avg_temp` as parameters**

Coefficient of correlation is equal to `0.7944540448`

**Scatter plot with `gd_avg_temp` as x-axis and `cd_avg_temp` as y-axis**
![Image](./image/scatter_guarulhos_city_data_year_c.png)

**Predicting value with Regression function**

Is possible use $8.19 \times ln(x) + 2.18$ where $x$ = `gd_avg_temp`

## Multiple cities - Add your favorite cities from around the globe to your visualization. What do you learn about them?

**Data extraction**
```SQL
-- Average temperatures for Rio de Janeiro city by year (ºC).
SELECT * FROM city_data
 WHERE city = 'Rio De Janeiro'
   AND avg_temp IS NOT NULL;

-- Average temperatures for Dublin city by year (ºC).
SELECT * FROM city_data
 WHERE city = 'Dublin'
   AND avg_temp IS NOT NULL;

-- Average temperatures for Dallas city by year (ºC).
SELECT * FROM city_data
 WHERE city = 'Dallas'
   AND avg_temp IS NOT NULL;
```

**Save data into `rio_de_janeiro_city_data.csv` file**

In [5]:
rio_de_janeiro_city_data = pd.read_csv("./data/rio_de_janeiro_city_data.csv")
print(rio_de_janeiro_city_data)

     year            city country  avg_temp
0    1832  Rio De Janeiro  Brazil     23.05
1    1833  Rio De Janeiro  Brazil     24.11
2    1834  Rio De Janeiro  Brazil     23.27
3    1835  Rio De Janeiro  Brazil     22.73
4    1836  Rio De Janeiro  Brazil     22.91
5    1837  Rio De Janeiro  Brazil     22.29
6    1838  Rio De Janeiro  Brazil     22.81
7    1839  Rio De Janeiro  Brazil     22.54
8    1840  Rio De Janeiro  Brazil     23.32
9    1841  Rio De Janeiro  Brazil     22.97
10   1842  Rio De Janeiro  Brazil     23.41
11   1843  Rio De Janeiro  Brazil     23.55
12   1851  Rio De Janeiro  Brazil     23.53
13   1852  Rio De Janeiro  Brazil     23.74
14   1853  Rio De Janeiro  Brazil     23.76
15   1854  Rio De Janeiro  Brazil     23.83
16   1855  Rio De Janeiro  Brazil     23.89
17   1856  Rio De Janeiro  Brazil     22.83
18   1857  Rio De Janeiro  Brazil     23.46
19   1858  Rio De Janeiro  Brazil     22.37
20   1859  Rio De Janeiro  Brazil     23.04
21   1860  Rio De Janeiro  Brazi

**Save data into `dublin_city_data.csv` file**

In [6]:
dublin_city_data = pd.read_csv("./data/dublin_city_data.csv")
print(dublin_city_data)

     year    city  country  avg_temp
0    1743  Dublin  Ireland      6.80
1    1744  Dublin  Ireland      9.28
2    1745  Dublin  Ireland      4.32
3    1750  Dublin  Ireland      9.32
4    1751  Dublin  Ireland      9.12
5    1752  Dublin  Ireland      6.28
6    1753  Dublin  Ireland      8.63
7    1754  Dublin  Ireland      8.59
8    1755  Dublin  Ireland      8.29
9    1756  Dublin  Ireland      8.75
10   1757  Dublin  Ireland      8.72
11   1758  Dublin  Ireland      7.83
12   1759  Dublin  Ireland      8.82
13   1760  Dublin  Ireland      8.14
14   1761  Dublin  Ireland      8.95
15   1762  Dublin  Ireland      8.32
16   1763  Dublin  Ireland      7.95
17   1764  Dublin  Ireland      8.26
18   1765  Dublin  Ireland      8.08
19   1766  Dublin  Ireland      8.24
20   1767  Dublin  Ireland      8.38
21   1768  Dublin  Ireland      8.25
22   1769  Dublin  Ireland      8.33
23   1770  Dublin  Ireland      8.08
24   1771  Dublin  Ireland      8.14
25   1772  Dublin  Ireland      8.22
2

**Save data into `dallas_city_data.csv` file**

In [7]:
dallas_city_data = pd.read_csv("./data/dallas_city_data.csv")
print(dallas_city_data)

     year    city        country  avg_temp
0    1820  Dallas  United States     16.88
1    1821  Dallas  United States     17.33
2    1822  Dallas  United States     17.87
3    1823  Dallas  United States     17.46
4    1824  Dallas  United States     17.90
5    1825  Dallas  United States     18.38
6    1826  Dallas  United States     17.93
7    1827  Dallas  United States     18.62
8    1828  Dallas  United States     18.26
9    1829  Dallas  United States     17.89
10   1830  Dallas  United States     18.68
11   1831  Dallas  United States     16.98
12   1832  Dallas  United States     17.81
13   1833  Dallas  United States     18.29
14   1834  Dallas  United States     18.50
15   1835  Dallas  United States     16.58
16   1836  Dallas  United States     16.67
17   1837  Dallas  United States     17.47
18   1838  Dallas  United States     16.54
19   1839  Dallas  United States     17.64
20   1840  Dallas  United States     17.62
21   1841  Dallas  United States     17.66
22   1842  

**Rio de Janeiro 10 years moving average**
![Image](./image/rio_de_janeiro_10yma_city_data_year_c.png)

**Dublin 10 years moving average**
![Image](./image/dublin_10yma_city_data_year_c.png)

**Dallas 10 years moving average**
![Image](./image/dallas_10yma_city_data_year_c.png)

All cities have the same temperature behavior: the average temperature close to nowaday is higher than the average of the past years.